src/goto-programs/goto_check.cpp

```diff
@@ -342,16 +343,39 @@ void goto_checkt::bounds_check(
 
   assert(is_array_type(t) || is_string_type(t) || is_vector_type(t));
 
-  // We can't check the upper bound of an infinite sized array
-  if(is_array_type(t) && to_array_type(t).size_is_infinite)
-    return;
-
-  const expr2tc &array_size =
+  expr2tc array_size =
     is_array_type(t) ? to_array_type(t).array_size
     : is_vector_type(t)
       ? to_vector_type(t).array_size
       : constant_int2tc(get_uint32_type(), to_string_type(t).get_length());
 
+  // Are we dealing with infinity/incomplete arrays?
+  if(
+    is_array_type(t) &&
+    (to_array_type(t).size_is_infinite ||
+     (to_constant_int2t(to_array_type(t).array_size).value == 0)))
```

Comment:
can we have an array member with size 0 inside a struct, that is not a fam?

Reply:
There are zero-length arrays: https://gcc.gnu.org/onlinedocs/gcc/Zero-Length.html

I didn't look too deep into it but the main difference seems to be that FAM allows static initialization of the array while zero-length does not.